feat: command auto-config

ematipico opened the pull request:
<!--
	Thanks for submitting a pull request!

	We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request.

	Once created, your PR will be automatically labeled according to changed files.

	Before submitting a pull request, please make sure the following is done:

	1. Format and resolve any lint errors: `./rome check --apply`
	2. Verify that tests pass: `./rome test`
	3. Ensure there are no TypeScript errors: `./node_modules/.bin/tsc`

	Learn more about contributing: https://github.com/rome/tools/blob/main/CONTRIBUTING.md
-->

## Summary

Closes #1112 
Fixes #1320 

This PR makes some refactors around the `init` command and it implements a new command called `auto-config`. Also made some changes and fixes around how the licenses are handled.

- the `init` command has been split in client and server command. The `init` command now asks for the type of configuration, type of indentation and size of indentation; this change is reflected to the generated config file and the `.editorconfig`. The command now doesn't do any lint check. The `init` command is successful only if there's not configuration file present inside the directory; The installation process now support `pnpm` too by checking its lock file;
- the `auto-config` command is split in client and server command. The server command does two different things pages on the flags passed to it. The command cannot be run inside a project that doesn't have a configuration (the diagnostic suggests to run `npx rome init`). The runs lint and inspects all licenses of the projects. All the diagnostics are then passed to the client which will parse them and run the relative `config set`/`config push` commands;


In order to retrieve the diagnostics of the licenses I had to low down the severity of the diagnostics generated by the codec of the manifest, otherwise  I wouldn't been able to collect those diagnostics. @sebmck is there a better way to collect these diagnostics during the `auto-config` while we run a strict check for the rest of the commands? 

I removed the hardcoded licenses we had in the code and we use the config to shut down the errors (`invalidLices`). A nice example of a working configuration! I had to manually add them.  

I also prepared some ground around the config handler, in particular `toml`. It's not ready yet but I put there some tests. At the moment nested objects are not handled at all.


Is there something else that we could add to `auto-config`?


What's missing:
- [x] tests for `init` and tests for `auto-config`
- [x] update documentation 



<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan

Run `./rome ci`

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output. -->

sebmck commented on internal/core/server/commands/autoConfig.ts:
```diff
@@ -0,0 +1,124 @@
+	checkVSC: boolean;
+}
+
+export type AutoConfig = RSERValue & {
```
Why `RSERValue &` the type of `AutoConfig` will widen.

ematipico replied:
Yeah that old code. Will clean it up

sebmck commented on internal/core/server/commands/autoConfig.ts:
```diff
@@ -0,0 +1,124 @@
+			return;
+		}
+
+		if (!currentProject.initialized) {
```
This should never happen after an `assertProject` call. `initialized` marks whether we're still crawling the project on disk. When calling `assertProject` we would wait on it if it was initiated by another call.

sebmck commented on internal/core/server/commands/config.ts:
```diff
@@ -76,9 +76,9 @@ async function runCommand(
 
 		reporter.success(
-			markup`${action === "push" ? "Adding" : "Setting"} <emphasis>${keyParts}</emphasis> to <emphasis>${JSON.stringify(
+			markup`${action === "push" ? "Adding" : "Setting"} <emphasis>${JSON.stringify(
```
Using `prettyFormat(value)` here instead would be more appropriate I think.

sebmck commented on internal/core/server/utils/getManifest.ts:
```diff
@@ -0,0 +1,24 @@
+import {json} from "@internal/codec-config";
+import {Server} from "@internal/core";
+
+export default async function getManifest(
```
I think you can already use `server.memoryFs.getOwnedManifest(path)` to do what this method does.

ematipico replied:
I was looking for it everywhere!! Gosh 😅

sebmck commented on internal/core/server/utils/retrieveConfigHandler.ts:
```diff
@@ -0,0 +1,27 @@
+import {ConfigHandler, ConfigType} from "@internal/codec-config/types";
+import {json, rjson} from "@internal/codec-config";
+
+export default function retrieveConfigHandler(
```
The `codec-config` package is probably a better place for this.

sebmck commented on internal/core/server/utils/updateConfig.ts:
```diff
@@ -0,0 +1,28 @@
+import {DeepPartial} from "@internal/typescript-helpers";
+
+interface UpdateConfig {
+	config: Partial<ProjectConfig>;
```
`ProjectConfig` isn't the appropriate type to use when we are dealing with the actual raw project config values as it includes a lot of normalized data. It's actually clearer to just use `RSERObject` or something instead.

sebmck reviewed (commented):
Nice! Really excited about this.

> In order to retrieve the diagnostics of the licenses I had to low down the severity of the diagnostics generated by the codec of the manifest, otherwise I wouldn't been able to collect those diagnostics. @sebmck is there a better way to collect these diagnostics during the `auto-config` while we run a strict check for the rest of the commands?

There isn't. There's probably something we can do about lowering the severity of manfiest validation errors for all third party dependencies. That way we could even relax a lot of the `loose` checks, or remove it entirely. I'll have to think about it.

ematipico commented:
@sebmck thank you for the review! I updated the PR with the changes you highlighted. Unfortunately integration tests are still turned off so there's not much I can do around it. I tested locally though.

I updated the documentation. I would be great if we could add a screenshot but I don't know how to do it.

sebmck commented:
🥳
